dot/parachain/overseer/overseer.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package overseer
+
+import "fmt"
+
+type Overseer struct {
+	subsystems        map[Subsystem]*Context
+	subsystemMessages map[Subsystem]chan any
+}
+
+type ExamlpeSender struct {
+	senderChan chan any
+}
+
+func (s *ExamlpeSender) SendMessage(msg any) error {
+	s.senderChan <- msg
+	return nil
+}
+
+func NewOverseer() *Overseer {
+	return &Overseer{
+		subsystems:        make(map[Subsystem]*Context),
+		subsystemMessages: make(map[Subsystem]chan any),
+	}
+}
+
+func (o *Overseer) RegisterSubsystem(subsystem Subsystem) {
+	// TODO: determine best buffer size
+	comChan := make(chan any, 5)
+	o.subsystemMessages[subsystem] = comChan
+	receiverChan := make(chan any, 5)
+	o.subsystems[subsystem] = &Context{
+		Sender:   &ExamlpeSender{senderChan: comChan},
+		Receiver: receiverChan,
+	}
+}
+
+func (o *Overseer) Start() {
+	// start subsystems
+	for subsystem, context := range o.subsystems {
+		go func(sub Subsystem, ctx *Context) {
+			err := sub.Run(ctx)
+			if err != nil {
+				// TODO: handle error
+				fmt.Printf("error running subsystem %v: %v\n", sub, err)
+			}
+		}(subsystem, context)
+	}
+
+	// wait for messages from subsystems
+	// TODO: this is a temporary solution, we will determine logic to handle different message types
+	for subsystem, recChan := range o.subsystemMessages {
+		go func(sub Subsystem, channel chan any) {
+			fmt.Printf("overseer waiting for messages from %v\n", sub)
+
+			for {
```

Comment:
We should add a return or break in this for loop to end this goroutine. Currently, it is a forever goroutine.

Reply:
Agree, we should pass to the `Start` method a `stop chan struct{}` argument that goroutines should listen for and whenever it triggers all goroutines should return. Also, I would suggest having a `sync.WaitGroup` in the `Overseer` struct and for each goroutine we increment 1 in the wait group

```go
type Overseer struct {
    wg sync.WaitGroup
    doneCh chan struct{}
    stopCh chan struct{}
}

// here stop channel is just a read-only channel which means that
func NewOverseer() *Overseer {
  return &Overseer{
    doneCh: make(chan struct{}),
    stopCh: make(chan struct{}),
  }
}

func (o *Overseer) Start() {
    for subsystem, context := range o.subsystems {
        o.wg.Add(1)
        go func(..., wg *sync.WaitGroup) {
            defer wg.Done()
            select {
            case <-o.stopCh:
               return // here you can shutdown gracefully the goroutine
            case ...
            }
        }
    }

    for subsystem, recChan := range o.subsystemMessages {
        o.wg.Add(1)
        go func(..., wg *sync.WaitGroup) {
            defer wg.Done()
            select {
            case <-o.stopCh:
               return // here you can shutdown gracefully the goroutine
            case ...
            }
        }
    }
}

func (o *Overseer) Stop() {
    if o.doneCh == nil {
        return nil // or return an error 
    }

    close(o.stopCh)
    // once you close the stop channel you should
    // wait for the goroutines to end until
    o.wg.Wait()

    // assign nil to stopCh prevents subsequent calls to
    // Stop method
    o.stopCh = nil  
}
```